DAOS-16982 csum: recalculate checksum on retrying

jxiong opened the pull request:
This PR fixes retry logic by actually recalculating the checksum; also it removes the code that incorrectly records nvme error.

This is a quick fix before we make an ultimate fix discussed here: https://daos-stack.slack.com/archives/C4SM0RZ54/p1738030213108609

Change-Id: Ib0287851fea4d125eecda48c5ccb3c73ed85b8f8
Signed-off-by: Jinshan Xiong <jinshanx@google.com>

### Before requesting gatekeeper:

* [ ] Two review approvals and any prior change requests have been resolved.
* [ ] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [ ] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [ ] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [ ] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate owners.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

jxiong commented:
I have already tested it by manually injecting failure, and I'm working on turning that into a unit test.
```
diff --git a/src/object/cli_obj.c b/src/object/cli_obj.c
index fdd9528a0..700abe77e 100644
--- a/src/object/cli_obj.c
+++ b/src/object/cli_obj.c
@@ -5141,6 +5141,7 @@ obj_csum_update(struct dc_object *obj, daos_obj_update_t *args, struct obj_auxi_
 		return 0;

 	if (obj_auxi->csum_retry) {
+	  	D_ERROR("recalculate csum error\n");
 		/* release old checksum result and prepare for new calculation */
 		daos_csummer_free_ic(obj->cob_co->dc_csummer, &obj_auxi->rw_args.iod_csums);
 	}
@@ -5156,6 +5157,7 @@ obj_csum_fetch(const struct dc_object *obj, daos_obj_fetch_t *args,
 	       struct obj_auxi_args *obj_auxi)
 {
 	if (obj_auxi->csum_retry) {
+	  	D_ERROR("recalculate csum error\n");
 		/* release old checksum result and prepare for new calculation */
 		daos_csummer_free_ic(obj->cob_co->dc_csummer, &obj_auxi->rw_args.iod_csums);
 	}
diff --git a/src/object/srv_obj.c b/src/object/srv_obj.c
index 26240540b..aae06ef61 100644
--- a/src/object/srv_obj.c
+++ b/src/object/srv_obj.c
@@ -1366,10 +1366,16 @@ obj_local_rw_internal(crt_rpc_t *rpc, struct obj_io_context *ioc, daos_iod_t *io
 		D_GOTO(out, rc = 0);
 	}

+	static int fail_count = 5;
+
 	rc = csum_verify_keys(ioc->ioc_coc->sc_csummer, &orw->orw_dkey,
 			      orw->orw_dkey_csum, &orw->orw_iod_array,
 			      &orw->orw_oid);
-	if (rc != 0) {
+	if (fail_count > 0) {
+	  	fail_count--;
+		rc = -DER_CSUM;
+	}
+	if (rc != 0 || fail_count > 0) {
 		D_ERROR(DF_C_UOID_DKEY"verify_keys error: "DF_RC"\n",
 			DP_C_UOID_DKEY(orw->orw_oid, &orw->orw_dkey),
 			DP_RC(rc));
```

jolivier23 commented on src/object/cli_obj.c:
```diff
@@ -5140,6 +5141,11 @@ obj_csum_update(struct dc_object *obj, daos_obj_update_t *args, struct obj_auxi_
 
+	if (obj_auxi->csum_retry) {
+		/* Release old checksum result and prepare for new calculation */
+		daos_csummer_free_ic(obj->cob_co->dc_csummer, &obj_auxi->rw_args.iod_csums);
```
I think we probably want to do this after a couple of retries

jxiong replied:
It's really easy to add but I wonder if that is indeed necessary, because cksum error is a rare event by itself.

How about revising it to:
```
if (obj_auxi->csum_retry && obj_auxi->csum_retry_cnt > 2) { ... }
```
would that work for you?

jolivier23 replied:
yeah

jolivier23 commented on src/object/cli_obj.c:
```diff
@@ -5140,6 +5141,11 @@ obj_csum_update(struct dc_object *obj, daos_obj_update_t *args, struct obj_auxi_
+
 	return dc_obj_csum_update(obj->cob_co->dc_csummer, obj->cob_co->dc_props,
 				  obj->cob_md.omd_id, args->dkey, args->iods, args->sgls, args->nr,
 				  obj_auxi->reasb_req.orr_singv_los, &obj_auxi->rw_args.dkey_csum,
```
In the case of the actual issue we saw, it was the dkey_csum that needs to be recalculated, is that happening here?

jxiong replied:
Yes if I read the code correctly because we release the previous calculation above.

daosbuild1 commented:
Test stage Build RPM on EL 8 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-15786/3/execution/node/344/log

daosbuild1 commented:
Test stage Build on EL 8 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-15786/3/execution/node/387/log

daosbuild1 commented:
Test stage Build RPM on EL 9 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-15786/3/execution/node/345/log

daosbuild1 commented:
Test stage Build on Leap 15.5 with Intel-C and TARGET_PREFIX completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-15786/3/execution/node/480/log

daosbuild1 commented:
Test stage Build DEB on Ubuntu 20.04 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-15786/3/execution/node/339/log

daosbuild1 commented:
Test stage Build on EL 8 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-15786/4/execution/node/375/log

daosbuild1 commented:
Test stage Build on Leap 15.5 with Intel-C and TARGET_PREFIX completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-15786/4/execution/node/373/log

daosbuild1 commented:
Test stage Build RPM on EL 9 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-15786/4/execution/node/345/log

daosbuild1 commented:
Test stage Build RPM on EL 8 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-15786/4/execution/node/338/log

daosbuild1 commented:
Test stage Build RPM on Leap 15.5 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-15786/4/execution/node/335/log

daosbuild1 commented:
Test stage Build DEB on Ubuntu 20.04 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-15786/4/execution/node/342/log

daosbuild1 commented:
Test stage Build on Leap 15.5 with Intel-C and TARGET_PREFIX completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-15786/5/execution/node/373/log

daosbuild1 commented:
Test stage Build RPM on EL 9 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-15786/5/execution/node/319/log

daosbuild1 commented:
Test stage Build RPM on EL 8 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-15786/5/execution/node/345/log

daosbuild1 commented:
Test stage Build RPM on Leap 15.5 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-15786/5/execution/node/342/log

daosbuild1 commented:
Test stage Build on Leap 15.5 with Intel-C and TARGET_PREFIX completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-15786/6/execution/node/374/log

daosbuild1 commented:
Test stage Build RPM on EL 9 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-15786/6/execution/node/371/log

daosbuild1 commented:
Test stage Build RPM on EL 8 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-15786/6/execution/node/356/log

daosbuild1 commented:
Test stage Build RPM on Leap 15.5 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-15786/6/execution/node/355/log

daosbuild1 commented:
Test stage Build DEB on Ubuntu 20.04 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-15786/6/execution/node/359/log

daosbuild1 commented:
Test stage Build on Leap 15.5 with Intel-C and TARGET_PREFIX completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-15786/7/execution/node/373/log

daosbuild1 commented:
Test stage Build RPM on EL 9 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-15786/7/execution/node/348/log

daosbuild1 commented:
Test stage Build RPM on EL 8 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-15786/7/execution/node/347/log

daosbuild1 commented:
Test stage Build RPM on Leap 15.5 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-15786/7/execution/node/346/log

daosbuild1 commented:
Test stage Build DEB on Ubuntu 20.04 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-15786/7/execution/node/336/log

wangdi1 commented on src/object/cli_obj.c:
```diff
@@ -4758,38 +4759,37 @@ obj_comp_cb(tse_task_t *task, void *data)
-					if (task->dt_result == -DER_CSUM)
+				if ((obj_auxi->opc == DAOS_OBJ_RPC_FETCH ||
+				     obj_auxi->opc == DAOS_OBJ_RPC_UPDATE) &&
+				    task->dt_result == -DER_CSUM) {
```
I thought we only want to retry for csum error on wire? but not for NVME fetch CSUM error? No?

jxiong replied:
Yeah that's correct, but it would lead to a much large change, and we also want to get a confirmation from the DAOS community to make sure it's fine to move that forward.

Therefore, our strategy here is to land a quick fix while waiting for it.

wangdi1 replied:
oh, ok

wangdi1 commented on src/object/cli_obj.c:
```diff
@@ -4758,38 +4759,37 @@ obj_comp_cb(tse_task_t *task, void *data)
+					 * go terribly wrong if checksum happened for 10 times in a
+					 * row therefore we stop trying anyways even though there
+					 * may exist more replicas. */
+					if (rw_arg->csum_retry_cnt < MAX_CSUM_RETRY) {
```
for fetch, it should be

if (rw_arg->csum_retry_cnt < max(MAX_CSUM_RETRY, obj_get_replicas(obj)) ?

wangdi1 replied:
ah, did not notice your comments.

daosbuild1 commented:
Test stage Build on Leap 15.5 with Intel-C and TARGET_PREFIX completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-15786/8/execution/node/507/log

daosbuild1 commented:
Test stage Build DEB on Ubuntu 20.04 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-15786/8/execution/node/332/log

daosbuild1 commented:
Test stage Unit Test on EL 8.8 completed with status UNSTABLE.  https://build.hpdd.intel.com/job/daos-stack/job/daos//view/change-requests/job/PR-15786/9/testReport/

daosbuild1 commented:
Test stage Unit Test with memcheck on EL 8.8 completed with status UNSTABLE.  https://build.hpdd.intel.com/job/daos-stack/job/daos//view/change-requests/job/PR-15786/9/testReport/

jolivier23 commented:
This PR really should be against master branch (and then backported)

jolivier23 commented:
unit tests failing, those are pretty simple to run locally.

need to mount a tmpfs at /mnt/daos and then run utils/utest.py

jxiong commented:
Close for now. Will be pushed to master first

jxiong commented:
@wangdi1 @liuxuezhao can you please take a look?

jolivier23 commented on src/object/srv_obj.c:
```diff
@@ -1371,11 +1372,8 @@ obj_local_rw_internal(crt_rpc_t *rpc, struct obj_io_context *ioc, daos_iod_t *io
-			DP_C_UOID_DKEY(orw->orw_oid, &orw->orw_dkey),
-			DP_RC(rc));
-		if (rc == -DER_CSUM)
-			obj_log_csum_err();
```
Perhaps we should fix this in a separate patch?

liuxuezhao commented on src/object/cli_obj.c:
```diff
@@ -4758,38 +4759,37 @@ obj_comp_cb(tse_task_t *task, void *data)
 				/* Retry fetch on alternative shard */
-				if (obj_auxi->opc == DAOS_OBJ_RPC_FETCH) {
-					if (task->dt_result == -DER_CSUM)
+				if ((obj_auxi->opc == DAOS_OBJ_RPC_FETCH ||
```
now if FETCH get DER_CSUM
1. for replica obj it will try other replica, if all replicas retried then will fail, see obj_retry_next_shard(), 
    so actually the "csum_retry_cnt < MAX_CSUM_RETRY" compare for fetch looks not very reasonable, because if #replicas < MAX_CSUM_RETRY the check is useless (alwasy true) , if #replicas > MAX_CSUM_RETRY it acutally has chance to sucess but the code will fail it.
2. for EC obj it will mark the shard as failed (obj_auxi_add_failed_tgt()) and do EC degraded fetch, so the "csum_retry_cnt < MAX_CSUM_RETRY" compare also not really useful, because if the retried times exceed number of parity shards it will fail.

The retry for UPDATE is easier just retry that times, the check is valid.

so looks to me that the change here make the code a kind of inconsistent with real behavior (for FETCH). Do you think is it fine? I'll leave a -1 first, thx

jxiong replied:
`csum_retry_cnt < MAX_CSUM_RETRY` will be applied to both read and write. For read, even though it may have more than 10 (`MAX_CSUM_RETRY` is set to 10 at this point) replicas, but if it returns the same error 10 times in a row, something may have gone terribly wrong. Either way, I think it would be reasonable to limit the rw RPC to retry only limited times. If there exists BUGs in our code, this will prevent it from unlimited trying, which we have seen this in production.

I actually have a comment below to explain when I did this. Or otherwise, please suggest a fix.

liuxuezhao replied:
I was thought that for FETCH need not consider the csum_retry_cnt as it with its own different control.
I thought again that as long as the MAX_CSUM_RETRY will not be changed smaller probably is fine.

liuxuezhao commented on src/object/cli_obj.c:
```diff
@@ -5140,6 +5140,12 @@ obj_csum_update(struct dc_object *obj, daos_obj_update_t *args, struct obj_auxi_
 	if (!obj_csum_dedup_candidate(&obj->cob_co->dc_props, args->iods, args->nr))
 		return 0;
 
+	if (obj_auxi->csum_retry) {
```
just confirm that will the original code cause mem leak?  (as it did not call the daos_csummer_free_ci() before).
and why only do the daos_csummer_free_ci for csum_retry, need not do it for other retry like nvme_io_err/tx_uncertain?

jxiong replied:
This is not about memory leak. The checksum computation code `dc_obj_csum_update()` will use `obj_auxi->rw_args.dkey_csum` and `obj_auxi->rw_args.iod_csums` to check if the checksum has been computed.

By freeing the previous result, it will cause the checksum to be recomputed, which is what we want for the sake of this fix

liuxuezhao replied:
ok, thanks for explanation. 

For "changing NVME checksum error to nvme_io_err", 
"For DER_NVME_IO error, it should try next replica if it has; otherwise, it should just send the RPC to the same replica again and hopefully network error would disappear."
that will need change quite some codes and make the logic more complicate.
It depends on the possibility of network transferring caused CSUM err, is it high or very rare, when it happens can it be corrected by retry the transferring to same target?
It looks to me that possibility (pure network caused csum err and fixed by re-transferring) should be very rare. On the other side, RPC (such as mercury) internally can verify the CSUM during transferring and will fail the underneath RPC.
So I'm wondering that possibly not worth to introduce much complexity for that looks-not-true assumption.

jolivier23 reviewed (commented):
Probably need to refresh with latest master, not sure what happened with the builds

jxiong committed:
DAOS-16982 csum: recalculate checksum on retrying

This PR fixes retry logic by actually recalculating the checksum; also
it removes the code that incorrectly records nvme error.

Run-GHA: true

Change-Id: Ib0287851fea4d125eecda48c5ccb3c73ed85b8f8
Signed-off-by: Jinshan Xiong <jinshanx@google.com>

jxiong commented:
> Probably need to refresh with latest master, not sure what happened with the builds

Done

kjacque commented:
Removed @daos-stack/go-owners  and @daos-stack/go-watchers  groups because there don't seem to be any Go changes in this PR. If that changes, feel free to re-add.

jxiong committed:
fix per comment

jxiong committed:
add sign-off

Signed-off-by: Jinshan Xiong <jinshanx@google.com>

jxiong commented:
@jolivier23 @wangdi1 @liuxuezhao can I get a review on this?

liuxuezhao commented:
> @jolivier23 @wangdi1 @liuxuezhao can I get a review on this?

quite some csum tests failed https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-15786/19/

liuxuezhao commented:
> > > @jolivier23 @wangdi1 @liuxuezhao can I get a review on this?
> > 
> > 
> > quite some csum tests failed https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-15786/19/
> 
> I'm not able to see this, can you post it here if possible?

[Test Hardware / Functional Hardware Medium Verbs Provider / FTEST_daos_test.DaosCoreTest-DAOS_Checksum.1. DAOS_CSUM01: simple update with server side verify](https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-15786/19/testReport/junit/FTEST_daos_test/DaosCoreTest-DAOS_Checksum/Test_Hardware___Functional_Hardware_Medium_Verbs_Provider___1__DAOS_CSUM01__simple_update_with_server_side_verify/)
[Test Hardware / Functional Hardware Medium Verbs Provider / FTEST_daos_test.DaosCoreTest-DAOS_Checksum.19. DAOS_CSUM06: Single Value Checksum](https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-15786/19/testReport/junit/FTEST_daos_test/DaosCoreTest-DAOS_Checksum/Test_Hardware___Functional_Hardware_Medium_Verbs_Provider___19__DAOS_CSUM06__Single_Value_Checksum/)
[Test Hardware / Functional Hardware Medium Verbs Provider / FTEST_daos_test.DaosCoreTest-DAOS_Checksum.20. DAOS_CSUM07: Mix of Single Value and Array values iods](https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-15786/19/testReport/junit/FTEST_daos_test/DaosCoreTest-DAOS_Checksum/Test_Hardware___Functional_Hardware_Medium_Verbs_Provider___20__DAOS_CSUM07__Mix_of_Single_Value_and_Array_values_iods/)
[Test Hardware / Functional Hardware Medium Verbs Provider / FTEST_daos_test.DaosCoreTest-DAOS_Checksum.21. DAOS_CSUM08: Update/Fetch A Key](https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-15786/19/testReport/junit/FTEST_daos_test/DaosCoreTest-DAOS_Checksum/Test_Hardware___Functional_Hardware_Medium_Verbs_Provider___21__DAOS_CSUM08__Update_Fetch_A_Key/)
[Test Hardware / Functional Hardware Medium Verbs Provider / FTEST_daos_test.DaosCoreTest-DAOS_Checksum.22. DAOS_CSUM09: Update/Fetch D Key](https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-15786/19/testReport/junit/FTEST_daos_test/DaosCoreTest-DAOS_Checksum/Test_Hardware___Functional_Hardware_Medium_Verbs_Provider___22__DAOS_CSUM09__Update_Fetch_D_Key/)
[Test Hardware / Functional Hardware Medium Verbs Provider / FTEST_daos_test.DaosCoreTest-DAOS_Checksum.23. DAOS_CSUM09.2: Update/Fetch D Key](https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-15786/19/testReport/junit/FTEST_daos_test/DaosCoreTest-DAOS_Checksum/Test_Hardware___Functional_Hardware_Medium_Verbs_Provider___23__DAOS_CSUM09_2__Update_Fetch_D_Key/)
[Test Hardware / Functional Hardware Medium Verbs Provider / FTEST_daos_test.DaosCoreTest-DAOS_Checksum.24. DAOS_CSUM10: Enumerate A Keys](https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-15786/19/testReport/junit/FTEST_daos_test/DaosCoreTest-DAOS_Checksum/Test_Hardware___Functional_Hardware_Medium_Verbs_Provider___24__DAOS_CSUM10__Enumerate_A_Keys/)
[Test Hardware / Functional Hardware Medium Verbs Provider / FTEST_daos_test.DaosCoreTest-DAOS_Checksum.25. DAOS_CSUM11: Enumerate D Keys](https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-15786/19/testReport/junit/FTEST_daos_test/DaosCoreTest-DAOS_Checksum/Test_Hardware___Functional_Hardware_Medium_Verbs_Provider___25__DAOS_CSUM11__Enumerate_D_Keys/)
[Test Hardware / Functional Hardware Medium Verbs Provider / FTEST_daos_test.DaosCoreTest-DAOS_Checksum.26. DAOS_CSUM12: Enumerate objects](https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-15786/19/testReport/junit/FTEST_daos_test/DaosCoreTest-DAOS_Checksum/Test_Hardware___Functional_Hardware_Medium_Verbs_Provider___26__DAOS_CSUM12__Enumerate_objects/)
[Test Hardware / Functional Hardware Medium Verbs Provider / FTEST_daos_test.DaosCoreTest-DAOS_Checksum.45. DAOS_EC_CSUM01: simple update with server side verify](https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-15786/19/testReport/junit/FTEST_daos_test/DaosCoreTest-DAOS_Checksum/Test_Hardware___Functional_Hardware_Medium_Verbs_Provider___45__DAOS_EC_CSUM01__simple_update_with_server_side_verify/)
